### ISSUE TITLE
Issue with entry.fields

### ISSUE THREAD

**Issue by jfsagasti** [contentful/contentful.swift#34]
Hello! I'm using the iOS SDK for Swift 2.3. 

I have an entry which has an array of Stores (another type of entries) linked inside a field with "stores" key. When I try to fetch that entry, I encountered a problem trying to get useful info inside the fields structure:

```
	func fetchArray(withIdentifier identifier: String,
	                key: String,
	                success: @escaping (String, [Entry]) -> (),
	                failure: @escaping (NSError) -> ()) {
		client.fetchEntry(identifier: identifier) { result in
			switch result {
			case let .success(entry):
                guard let entries = entry.fields[key] as? [Entry] else { return failure(ContentfulError.keyMismatch(key) as NSError) } // Problem here because cast always fails
				success(entry.fields["description"] as? String ?? "", entries)
				
			case let .error(error):
				failure(error as NSError)
			}
		}
	}
```
The problem is in the guard clausule: entry.fields[key] as? [Entry] always fails because  fields[key] is always of type Any. I've tried to convert it to other objects without success. And I cannot parse it. This is also happening in the Swift 3 SDK. However, if I do exactly the same but with the Objective-C SDK, the cast to [CDAEntry] doesn't fail:

```
    func fetchArray(withIdentifier identifier: String,
                    key: String,
                    success: @escaping (String, [CDAEntry]) -> (),
                    failure: @escaping (NSError) -> ()) {
        client.fetchEntry(withIdentifier: identifier,
                          success: { (response: CDAResponse, entry: CDAEntry) in
                            guard let entries = entry.fields[key] as? [CDAEntry] else { return failure(ContentfulError()) }
                            success(entry.fields["description"] as! String, entries)
        }) { (response: CDAResponse?, error: Error) in
            failure(error as NSError)
        }
    }
```

Another interesting thing is that, using the Objective-C SDK, the linked entries (stores) are complete with all the Store model info. But in the Swift SDK, the Any object received only contains partial info (ids and so. I was able to know this because I printed .fields[key]  in the debugger).

 What is happening here? Is there really an issue or am I missing something? 
Thanks in advance!

**Comment by loudmouth** [comment]
Hi @jfsagasti, thanks for the report. Before I investigate further, I would love to get one more piece of information.

You mentioned that you are using Swift 2.3, but the code you posted above is Swift 3 code. Can you confirm which version of the SDK you are using? The Swift 2.3 SDK is on a separate maintenance branch at this point as Xcode 8.3+ will not support Swift 2.

**Comment by jfsagasti** [comment]
Hi @loudmouth! Thanks and sorry for the oversight. I've tested with both 2.3 and 3 SDKs (I pasted the 3 example in the end). Both gives the same problematic.

Here is the 2.3 version function:

```
    func fetchArray(withIdentifier identifier: String,
                                   key: String,
                                   success: ([Entry]) -> (),
                                   failure: (NSError) -> ()) {
        client.fetchEntry(identifier) { result in
            switch result {
            case let .Success(entry):
                guard let entries = entry.fields[key] as? [Entry] else { return failure(ContentfulError.keyMismatch(key) as NSError) }
                success(entries)
				
            case let .Error(error):
                failure(error as NSError)
            }
        }
    }
```

**Comment by loudmouth** [comment]
Perfect, thank you. I'll investigate immediately and will give you an update tomorrow.

In the meantime, the Objective-C SDK has more thorough test coverage, and if it suits your needs, I recommend using it.

**Comment by loudmouth** [comment]
@jfsagasti are you using Swift 2.3 though for the project you are currently working on? I would love to focus my efforts on `master` (Swift 3) if possible

**Comment by loudmouth** [comment]
@jfsagasti I would recommend using the objective-c SDK then. While I will fix this bug either way, I have received other tickets in recent months about the Swift SDK, and I have a much higher level of confidence in the older, and more battle tested Objective-C counterpart at this point.

I recommend using the latest version of the Obj-C SDK at the HEAD of `master` as it supports AF-Networking 3 and drops iOS 7. 

with Cocoapods you would add this to your Podfile.

```ruby
 pod 'ContentfulDeliveryAPI', :git => 'https://github.com/contentful/contentful.objc', :branch => 'master'
```

Carthage would be

github "contentful/contentful.objc" "master"

**Comment by jfsagasti** [comment]
Thanks a lot! The Obj-C seemed more robust at this point to me and I was just about to ask you for your recommendation, given our current state in the project.

**Comment by loudmouth** [comment]
Hi @jfsagasti, though I still recommend using the Objective-C SDK, I wanted to follow up and let you know that I have pushed a branch (see #35) with a fix to be merged with Swift 3.

The issue was that the Swift SDK was fetching an entry with the `entries/entry_id` endpoint which does _not_ automatically resolve includes. In contrast, the `fetchEntryWithIdentifier` method in  contentful.objc fetches a single entry by matching `sys.id`. This endpoint _does_ resolve 1 level of includes and will therefore return 1 level of linked entries/assets. (Note that links of links will not be resolved unless you specify `["include" : X]` where `X > 1`)

If you are curious to learn more about the way the Contentful API handles Linked resources, see the [Links Documentation](https://www.contentful.com/developers/docs/concepts/links/).

**Comment by jfsagasti** [comment]
@loudmouth so 2.3 branch is outdated and won't be maintained anymore? Just to know, because this project still uses AFN 2.x and I need to discuss with the team the problematic until we migrate to Swift 3.

**Comment by loudmouth** [comment]
If you are still using AF-Networking 2, then you can use `v1.10.4` of the Objective-C SDK. It has no difference feature-wise from the yet to be published version of the SDK at the HEAD.

I would be happy to implement the fix for you on branch `Swift-2.3` as it's a minimal amount of code. Generally though, the syntax differences in Swift 2 and 3 are great enough that cherry picking commits from the Swift 3 branch to Swift 2 and vise-versa is a lot of extra work. Given that Xcode will be dropping Swift 2 support as soon as either a version > 8.2 comes out, the extra work to maintain Swift 2 is not worth the reward.

**Comment by loudmouth** [comment]
Again though, in my opinion, contentful.swift is not quite production ready. I highly recommend just sticking with Objective-C for now.
